synadm/api.py

```diff
@@ -1139,3 +1139,44 @@ def user_shadow_ban(self, user_id, unban):
         else:
             method = "post"
         return self.query(method, f"v1/users/{user_id}/shadow_ban")
+
+    def notice_send(self, receivers, content_plain, content_html, paginate):
+        """ Send server notices.
+
+        Args:
+            receivers: Target(s) of the notice. Either localpart or regular
+                expression matching localparts.
+            content_plain: Unformatted text of the notice.
+            content_html: HTML-formatted text of the notice.
+        """
+        data = {
+            "user_id": "",
+            "content": {
+                "msgtype": "m.text",
+                "body": content_plain,
+                "format": "org.matrix.custom.html",
+                "formatted_body": content_html
+            }
+        }
+
+        # A regular expression was supplied to match receivers.
+        if receivers[:1] == '^':
```

Comment:
In my experience, you use a `^` to indicate that you only want matches in a string that are found right at the beginning of that string, so you don't do any potentially incorrect multi-matching. 

But here the point is that, as you said, we can also have regex that doesn't use `^`. You are right that a distinction must be made here, but I find it not so easy to determine this on the basis of one or more characters, since these characters could potentially be part of a regex again. Spontaneously I would choose special characters, which are not allowed to be part of a MXID according to the matrix specification and at the same time can't be used for matching the allowed characters. However, I can't think of any character or character chain that would fulfill this condition.

Alternatively, a simple solution would be to add a flag for regex, with which the user can then explicitly indicate that the "search string" should be interpreted as regex. In fact I think this solution would be straight forward and one of the best. Let me know what you think about this proposal.

Reply:
@JOJ0 You're right, now I get what you mean.

@Ascurius I like your solution best, it's the most explicit and thus safe option to use imo. It makes the commands a little more verbose, but I think that's negligible wtih a short hand of the flag.

Maybe it would make sense to check for invalid characters in the `generate_mxid` function (in the else case where the mxid is constructed from a localpart)? So in case someone passes a regex without the flag, the function won't try to use the regex as a localpart... This could of course also be done in the CLI part, but I think hardening lib functions against arbitrary input is always a good idea. Let me know whether that makes sense :)

Reply:
My approach/suggestion still would be (as discussed in #synadm:peek-a-boo.at) to immitate what Ansible dev's did: Use the tilde character (~) as a prefix for regex, since it doesnt have any special meaning in any regex dialect. https://docs.ansible.com/ansible/2.8/user_guide/intro_patterns.html

On the other hand I always welcome an explicit and straight forward approach like using an option. You decide @a-0-dev what you prefer. Both perfectly fine solutions IMO.

Reply:
> Maybe it would make sense to check for invalid characters in the `generate_mxid` function (in the else case where the mxid is constructed from a localpart)? So in case someone passes a regex without the flag, the function won't try to use the regex as a localpart... This could of course also be done in the CLI part, but I think hardening lib functions against arbitrary input is always a good idea. Let me know whether that makes sense :)
> 

Yes very good idea, we should prevent that. I would like to have this change in a separate PR though. If you find the time that would be great, well, if not I'm being lazy and would also accept it within this PR ;-)😅

Reply:
I'll do this in a separate branch, but they should be accepted simultaneously because only having this PR merged without making `generate_mxid` more robust is prone to cause issues